src/source/avro.rs

```diff
@@ -134,47 +127,65 @@ impl TCPStreamHandler for AvroStreamHandler {
                             streaming = false;
                             break;
                         }
-                        _stream_token => match self.handle_avro_payload(
-                            chans.clone(),
-                            tags,
-                            &mut reader,
-                        ) {
-                            Ok(maybe_id) => {
-                                trace!("Avro payloads processed successfully.");
-                                AVRO_PAYLOAD_SUCCESS_SUM
-                                    .fetch_add(1, Ordering::Relaxed);
-
-                                if maybe_id.is_some() {
-                                    let mut resp = Cursor::new(Vec::new());
-                                    resp.write_u64::<BigEndian>(maybe_id.unwrap())
-                                        .expect("Failed to write response id!");
-                                    stream
-                                        .write_all(resp.get_ref())
-                                        .expect("Failed to write response!");
+                        _stream_token => {
+                            while streaming {
+                                match reader.read() {
+                                    Ok(payload) => {
+                                        let handle_res = self.handle_avro_payload(chans.clone(),tags, payload);
+                                        if handle_res.is_err() {
+                                            streaming = false;
+                                            break;
+                                        }
+
+                                        trace!("Avro payloads processed successfully.");
+                                            AVRO_PAYLOAD_SUCCESS_SUM
+                                            .fetch_add(1, Ordering::Relaxed);
+
+                                        let maybe_id = handle_res.unwrap();
+                                        if let Some(id) = maybe_id {
+                                            let mut resp = Cursor::new(Vec::new());
+                                            resp.write_u64::<BigEndian>(id)
+                                                    .expect("Failed to write response id!");
+                                            stream
+                                                .write_all(resp.get_ref())
+                                                .expect("Failed to write response!");
+                                        }
+                                    }
+
+                                    Err(PayloadErr::WouldBlock) => {
+                                        //Not enough data on the wire.  Try again later.
```

Comment:
Worth tracking this with internal telemetry?

Reply:
Naw, I don't think so.

Par for the nonblocking stream course.